packages/engine/paima-funnel/src/funnels/carp/funnel.ts

```diff
@@ -202,88 +205,253 @@ export class CarpFunnel extends BaseFunnel implements ChainFunnel {
-        ConfigNetworkType.CARDANO,
-        arg.from,
-        arg.to,
-        carpEvents
-      );
+                const cursor = cursors && cursors[extension.cdeId];
 
-      if (grouped.length > 0) {
-        data[this.chainName] = grouped;
-      }
+                const data = getCdeTransferData(
+                  this.carpUrl,
+                  extension,
+                  extension.startSlot,
+                  Math.min(startingSlot, extension.stopSlot || startingSlot),
+                  slot => slot,
+                  true,
+                  stableBlock.block.hash,
+                  cursor && cursor.kind === 'paginationCursor'
+                    ? (JSON.parse(cursor.cursor) as BlockTxPair)
+                    : undefined,
+                  this.config.paginationLimit
+                ).then(datums => {
+                  // we are providing the entire indexed range, so if carp
+                  // returns nothing we know the presync is finished for this
+                  // CDE.
+                  const finished =
+                    datums.length === 0 || datums.length < this.config.paginationLimit;
+
+                  cache.updateCursor(extension.cdeId, {
+                    kind: 'paginationCursor',
+                    cursor: datums[datums.length - 1]
+                      ? datums[datums.length - 1].paginationCursor.cursor
+                      : '',
+                    finished,
+                  });
+
+                  if (datums.length > 0) {
+                    datums[datums.length - 1].paginationCursor.finished = finished;
+                  }
 
-      return data;
-    } else {
-      const data = await basePromise;
+                  return datums;
+                });
 
-      if (arg) {
-        data[this.chainName] = FUNNEL_PRESYNC_FINISHED;
-      }
+                return data.then(data => ({
+                  cdeId: extension.cdeId,
+                  cdeType: extension.cdeType,
+                  data,
+                }));
+              }
+              default:
+                throw new Error(`[carp funnel] unhandled extension: ${extension.cdeType}`);
+            }
+          })
+      ),
+      basePromise,
+    ]);
+
+    const list: CardanoPresyncChainData[] = [];
+
+    for (const events of carpEvents) {
+      if (events.cdeType === ChainDataExtensionType.CardanoTransfer) {
+        for (const event of events.data || []) {
+          list.push({
+            extensionDatums: [event],
+            network: this.chainName,
+            networkType: ConfigNetworkType.CARDANO,
+            carpCursor: {
+              kind: 'paginationCursor',
+              cdeId: event.cdeId,
+              cursor: event.paginationCursor.cursor,
+              finished: event.paginationCursor.finished,
+            },
+          });
+        }
+      } else {
+        // handle the cde's that are still on slot range based 'pagination' (not
+        // really pagination, but emulated)
```

Comment:
do we still need this given we're merging all the Carp PRs?

Reply:
oh nvm I just noticed it's deleted by the follow-up PR

Reply:
we can merge that PR into this one if it makes it easier to review, though. It would also make easier to update the carp client version once released, since that one will contain the changes for all of this. Because if we make a release of carp client version now with the latest commit, we can't put it here since it won't compile because of the api changes.